tokio-util/src/sync/task_tracker.rs

```diff
@@ -0,0 +1,591 @@
+///             tokio::select! {
+///                 () = background_task(i) => {
+///                     println!("Task {} exiting normally.", i);
+///                 },
+///                 () = token.cancelled() => {
+///                     // Do some cleanup before we really exit.
+///                     time::sleep(Duration::from_millis(50)).await;
+///                     println!("Task {} finished cleanup.", i);
+///                 },
+///             }
+///         });
+///     }
+///
+///     // Spawn a background task that will send the shutdown signal.
+///     {
+///         let tracker = tracker.clone();
+///         tokio::spawn(async move {
+///             // Normally you would use something like ctrl-c instead of
+///             // sleeping.
+///             time::sleep(Duration::from_secs(2)).await;
+///             tracker.close();
+///             token.cancel();
+///         });
+///     }
+///
+///     // Wait for all tasks to exit.
+///     tracker.wait().await;
+///
+///     println!("All tasks have exited now.");
+/// }
+/// ```
+///
+/// [`CancellationToken`]: crate::sync::CancellationToken
+/// [`JoinHandle::is_finished`]: tokio::task::JoinHandle::is_finished
+/// [`JoinSet`]: tokio::task::JoinSet
+/// [`close`]: Self::close
+/// [`join_next`]: tokio::task::JoinSet::join_next
+/// [`wait`]: Self::wait
+/// [graceful shutdown]: https://tokio.rs/tokio/topics/shutdown
+#[derive(Clone)]
+pub struct TaskTracker {
+    inner: Arc<TaskTrackerInner>,
+}
+
+/// Represents a task tracked by a [`TaskTracker`].
+#[must_use]
+#[derive(Debug)]
+pub struct TaskTrackerToken {
+    task_tracker: TaskTracker,
+}
+
+struct TaskTrackerInner {
+    /// Keeps track of the state.
+    ///
+    /// The lowest bit is whether the task tracker is closed.
+    ///
+    /// The rest of the bits count the number of tracked tasks.
+    state: AtomicUsize,
+    /// Used to notify when the last task exits.
+    on_last_exit: Notify,
+}
+
+pin_project! {
+    /// A future that is tracked as a task by a [`TaskTracker`].
+    ///
+    /// The associated [`TaskTracker`] cannot complete until this future is dropped.
+    ///
+    /// This future is returned by [`TaskTracker::track_future`].
+    #[must_use = "futures do nothing unless polled"]
+    pub struct TrackedFuture<F> {
+        #[pin]
+        future: F,
+        token: TaskTrackerToken,
+    }
+}
+
+pin_project! {
+    /// A future that completes when the [`TaskTracker`] is empty and closed.
+    ///
+    /// This future is returned by [`TaskTracker::wait`].
+    #[must_use = "futures do nothing unless polled"]
+    pub struct TaskTrackerWaitFuture<'a> {
+        #[pin]
+        future: Notified<'a>,
+        inner: Option<&'a TaskTrackerInner>,
+    }
+}
+
+impl TaskTrackerInner {
+    #[inline]
+    fn new() -> Self {
+        Self {
+            state: AtomicUsize::new(0),
+            on_last_exit: Notify::new(),
+        }
+    }
+
+    #[inline]
+    fn is_done(&self) -> bool {
```

Comment:
naming nit: since the function that sets this is called `set_closed`, it seems like it would be more consistent for this to be `is_closed` rather than `is_done`?

Reply:
If you want another name, it would probably be `is_closed_and_empty`, since it also checks that the refcount is zero.